File: backend/src/main/java/com/smvita/computerseekho/controller/BatchAlbumController.java
package com.smvita.computerseekho.controller;

import com.smvita.computerseekho.dto.BatchAlbumDto;
import com.smvita.computerseekho.dto.BatchAlbumImageDto;
import com.smvita.computerseekho.service.BatchAlbumService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

/**
 * Nested under /batches because the album has no independent identity —
 * one per batch, and it lives inside the Batch Management screen (KB §9.5).
 */
@RestController
@RequestMapping("/batches/{batchId}/album")
@RequiredArgsConstructor
public class BatchAlbumController {

    private final BatchAlbumService batchAlbumService;

    @GetMapping
    public BatchAlbumDto find(@PathVariable Integer batchId) {
        return batchAlbumService.findByBatch(batchId);
    }

    @PostMapping("/images")
    public BatchAlbumDto addImage(@PathVariable Integer batchId,
                                  @Valid @RequestBody BatchAlbumImageDto image) {
        return batchAlbumService.addImage(batchId, image);
    }

    @PutMapping("/cover/{imageId}")
    public BatchAlbumDto setCover(@PathVariable Integer batchId, @PathVariable Integer imageId) {
        return batchAlbumService.setCover(batchId, imageId);
    }

    @DeleteMapping("/images/{imageId}")
    public BatchAlbumDto removeImage(@PathVariable Integer batchId, @PathVariable Integer imageId) {
        return batchAlbumService.removeImage(batchId, imageId);
    }

    @PutMapping
    public BatchAlbumDto updateDetails(@PathVariable Integer batchId,
                                       @RequestParam(required = false) String title,
                                       @RequestParam(required = false) String description) {
        return batchAlbumService.updateDetails(batchId, title, description);
    }
}

File: dotnet_backend/backend/Controllers/BatchAlbumController.cs
using ComputerSeekho.DTO;
using ComputerSeekho.Service;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ComputerSeekho.Controllers;

/// <summary>
/// Nested under /batches because an album has no independent identity —
/// one per batch, and it lives inside the Batch Management screen.
/// </summary>
[ApiController]
[Route("batches/{batchId:int}/album")]
[Authorize(Roles = "Admin,Manager")]
public class BatchAlbumController : ControllerBase
{
    private readonly IBatchAlbumService _service;

    public BatchAlbumController(IBatchAlbumService service) => _service = service;

    /// <summary>Anonymous, because the public Batch Album page reads it.</summary>
    [HttpGet]
    [AllowAnonymous]
    public async Task<ActionResult<BatchAlbumDto>> Get(int batchId, CancellationToken ct) =>
        Ok(await _service.GetOrCreateForBatchAsync(batchId, ct));

    [HttpPost("images")]
    public async Task<ActionResult<BatchAlbumDto>> AddImage(
        int batchId, BatchAlbumImageRequest request, CancellationToken ct) =>
        Ok(await _service.AddImageAsync(batchId, request, User.StaffId(), ct));

    [HttpPut("cover/{imageId:int}")]
    public async Task<ActionResult<BatchAlbumDto>> SetCover(int batchId, int imageId, CancellationToken ct) =>
        Ok(await _service.SetCoverAsync(batchId, imageId, ct));

    [HttpDelete("images/{imageId:int}")]
    public async Task<ActionResult<BatchAlbumDto>> RemoveImage(int batchId, int imageId, CancellationToken ct) =>
        Ok(await _service.RemoveImageAsync(batchId, imageId, ct));
}

/// <summary>
/// The public Campus Life strip.
///
/// A separate controller because the route is not nested under a batch —
/// it lists every album that has something to show.
/// </summary>
[ApiController]
[Route("batch-albums")]
public class BatchAlbumListController : ControllerBase
{
    private readonly IBatchAlbumService _service;

    public BatchAlbumListController(IBatchAlbumService service) => _service = service;

    [HttpGet]
    [AllowAnonymous]
    public async Task<ActionResult<IEnumerable<BatchAlbumSummaryDto>>> GetPublished(CancellationToken ct) =>
        Ok(await _service.GetPublishedAsync(ct));
}
